accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,1041 @@
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs if the inline class does not declare any
+instance variables, and if it declares two or more instance
+variables. Let `id` be the name of unique instance variable that it
+declares. The declaration of `id` must have the modifier `final`, and it
+can not have the modifier `late`; otherwise a compile-time error
+occurs.
+
+The _name of the representation_ in an inline class declaration is the
+name `id` of the unique final instance variable that it declares, and
+the _type of the representation_ is the declared type of `id`.
+
+A compile-time error occurs if an inline class declaration declares an
+abstract member.
+
+*There are no special rules for static members in inline classes. They
+can be declared and called or torn off as usual, e.g.,
+`Inline.myStaticMethod(42)`.*
+
+
+## Inline method invocations
+
+This document needs to refer to inline method invocations including
+each part that determines the static analysis and semantics of this
+invocation, so we will use a standardized phrase and talk about:
+An invocation of the inline member `m` on the receiver `e`
+according to the inline type `V` and with the actual type arguments
+<code>T<sub>1</sub>, ..., T<sub>s</sub></code>.
+
+In the case where `m` is a method, the invocation could be an inline
+member property extraction (*a tear-off*) or a method invocation,
+and in the latter case there will also be an `<argumentPart>`,
+optionally passing some actual type arguments, and (non-optionally)
+passing an actual argument list.
+
+The case where `m` is a setter is syntactically different, but the
+treatment is the same as with a method accepting a single argument
+(*so we will not specify that case explicitly*).
+
+*We need to mention all these elements together, because each of them
+plays a role in the static analysis and the dynamic semantics of the
+invocation.  There is no syntactic representation in the language for
+this concept, because the same inline method invocation can have many
+different syntactic forms, and both `V` and <code>T<sub>1</sub>, ...,
+T<sub>s</sub></code> are implicit in the actual syntax.*
+
+
+### Static Analysis of an Inline Member Invocation
+
+We need to introduce a concept that is similar to existing concepts
+for regular classes.
+
+We say that an inline class `V` _has_ a member named `n` in the case
+where `V` declares a member named `n`, and in the case where `V` has
+no such declaration, but `V` has a superinterface `Vs` that has a
+member named `n`. In both cases,
+_the member declaration named `n` that `V` has_ is said declaration.
+
+*This definition is unambiguous for an inline class that has no
```

Comment:
Mmm, not entirely right?  If `A` and `B` declare `foo`, `C` implements `A` and `B` and re-defines `foo`, and `D` implements `C`, then there are three such `Vs`.

Reply:
Added the word 'direct' just before 'superinterface'.